feat: implement stream api

mikewurtz opened the pull request:
This update adds support for streaming task output (stdout and stderr) via a new StreamTaskOutput API.

Summary
- Introduced a TaskWriter component that buffers combined stdout/stderr output from each task.
- Each connected client maintains a read offset into the shared output buffer and receives data in real time.
- Streaming is designed to support multiple concurrent clients. A slow client will not block others.
- Added integration tests to validate stream behavior, including full task output, mixed stdout/stderr, and client-side cancellations.

Additional Changes
- Introduced a timeout around grpcServer.GracefulStop() to avoid indefinite blocking during shutdown, particularly when long-lived streaming RPCs are active.

mikewurtz committed:
feat: Streaming API initial commit

mikewurtz committed:
fix: further improvements and error handling

mikewurtz committed:
fix: improve streaming by having separate mutexes

for output and the condition

mikewurtz committed:
chore: add a larger server shutdown and more tests

mikewurtz committed:
fix: use single RWmutex for cond

mikewurtz committed:
chore: run gofmt, add clarifying comments

rosstimothy commented on internal/grpc/client/manager.go:
```diff
@@ -73,13 +75,21 @@ func (m *Manager) StreamTaskOutput(ctx context.Context, taskID string) error {
-		return fmt.Errorf("error receiving from stream: %w", err)
+	for {
+		resp, err := stream.Recv()
+		if err == io.EOF {
```
Suggestion: prefer errors.Is
```suggestion
		if errors.Is(err, io.EOF) {
```

rosstimothy commented on internal/task/manager/stream.go:
```diff
@@ -0,0 +1,67 @@
+	var offset int64
+	for {
+		data, newOffset, err := taskObj.ReadOutput(mergedCtx, offset)
+		if err == io.EOF {
```
Suggestion: prefer errors.Is
```suggestion
		if errors.Is(err, io.EOF) {
```

rosstimothy commented on internal/task/manager/taskwriter.go:
```diff
@@ -0,0 +1,95 @@
+// Close closes the task writer and wakes up any waiting readers
+func (tw *TaskWriter) Close() {
+	tw.mu.Lock()
+	close(tw.done)
```
What would happen if Close was called more than once?

mikewurtz replied:
We would hit a panic. I was thinking this should only be called once as there is only one goroutine calling it but being defensive here is probably a better strategy. Will get this updated

mikewurtz committed:
Apply suggestions from code review

Co-authored-by: rosstimothy <39066650+rosstimothy@users.noreply.github.com>

mikewurtz committed:
fix: respond to feedback

make taskWriter.Close() more defensive

tigrato commented on internal/grpc/server/task_manager.go:
```diff
@@ -89,7 +89,28 @@ func checkAuthorization(caller string, taskObj *taskmanager.Task) error {
+		return stream.Send(&pb.StreamTaskOutputResponse{Output: data})
+	}
+
+	err = s.taskManager.StreamTaskOutput(stream.Context(), req.TaskId, writer)
```
What happens if the client closes the connection and no-new data is written to the output since then? 
Will the session terminate, will the goroutine become locked until a new write occurs?

mikewurtz replied:
As the code stands now, the goroutine is locked until a new write occurs so the goroutines could pile up here if we had a bunch of clients that cancelled their context on a job that doesn't produce much output. I will add a recheck interval. Great catch @tigrato!

tigrato replied:
What if you wake up the sync.Cond when a client wants to leave. Check context.AfterFunc .

That will also allow you to rewrite the streamer as an io.ReadCloser interface

mikewurtz replied:
This looks like a much better approach then having to wake up and poll constantly!

tigrato replied:
When I said I wanted to transform the streamer into an `io.ReadCloser`, I meant refactoring the entire call:

```go
err = s.taskManager.StreamTaskOutput(stream.Context(), req.TaskId, writer)
```

into something that behaves like this:

```go
jobStreamer, err := s.taskManager.StreamTaskOutput(ctx, req.TaskId)

context.AfterFunc(ctx, func() { jobStreamer.Close() })

for {
    n, err := jobStreamer.Read(buf)
    if errors.Is(err, io.EOF) {
        // ...
    } else if err != nil {
        // ...
    }

    err := stream.Send(&pb.StreamTaskOutputResponse{
        Output: slices.Clone(buf[:n]),
    })
    // ...
}
```

This design allows the caller to treat the result as a regular `io.Reader`, making it compatible with utilities like `io.Copy(os.Stdout, jobStreamer)` or `io.ReadAll(jobStreamer)` for more flexible consumption.

mikewurtz committed:
fix: add polling interval to clean up streams

clean up streams that are stuck on a cond.Wait() but the client
has disconnected

mikewurtz committed:
fix: use a context.Afterfunc and impl readcloser

timothyb89 commented on internal/task/manager/start.go:
```diff
@@ -79,7 +84,7 @@ func (tm *TaskManager) StartTask(ctx context.Context, command string, args []str
 	cgroupFd.Close()
```
nit: looks like there's an unhandled error here

(there's also one up a bit that I can't leave a comment on)

mikewurtz replied:
Good find, I guess the default golangci-lint settings do not pick up this error.

timothyb89 reviewed (commented):
Testing well on my machine so far, just noting a small nit

mikewurtz committed:
fix: fix unhandled Close() errors

also fix minor bug with setting system as termination source

rosstimothy commented on internal/task/manager/stream.go:
```diff
@@ -0,0 +1,79 @@
+
+// mergeCancelContexts merges two contexts and cancels when either is done
+func mergeCancelContexts(a, b context.Context) (context.Context, context.CancelFunc) {
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		select {
+		case <-a.Done():
+		case <-b.Done():
+		}
+		cancel()
+	}()
+	return ctx, cancel
```
Food for thought, but not something I expect you to address here: I think you could probably achieve the same result without the additional goroutine if you implemented a custom context.Context.

mikewurtz replied:
I think you are probably right, given that today is the last day of the challenge I will hold off on changing this.

rosstimothy commented on internal/task/manager/stream.go:
```diff
@@ -0,0 +1,79 @@
+		}
+
+		if n > 0 {
+			if err := writer(buf[:n]); err != nil {
```
What ownership guarantees does gRPC make about data sent over a stream?

mikewurtz replied:
This should probably be a copy

mikewurtz committed:
fix: address PR feedback

switch how streaming works

timothyb89 commented on internal/grpc/server/task_manager.go:
```diff
@@ -89,7 +93,47 @@ func checkAuthorization(caller string, taskObj *taskmanager.Task) error {
+	if err != nil {
+		return task.TaskErrorToGRPC(err)
+	}
+	context.AfterFunc(stream.Context(), func() { jobStreamer.Close() })
```
one final nit for the road, but looks like another unhandled `error` here

mikewurtz committed:
fix: fix up unhandled error
